src/core/mempool.rs

```diff
@@ -399,34 +399,67 @@ impl MemPoolDB {
         })
     }
 
-    fn walk(
+    /// Gets the lowest ancestor block(where `height > curr_tip_height && height <= max_tip_height`)
+    /// that is in the mempool of the "max" block (identified by `max_tip_block_hash` and
+    /// `max_tip_consensus_hash`). If found, this function returns a `MemPoolWalkResult::Chainstate`
+    /// enum along with identifying information about the ancestor block: its height
+    /// (`next_height`), its consensus hash, its block hash, and the earliest timestamp of the
+    /// transactions in it that are also in the mempool.
+    ///
```

Comment:
I have some questions about the approach, but there could be some constraints that I'm not seeing - I'm not that familiar with this part of the codebase.
Instead of reversing the walk, I wonder if we could / should decouple things, and split the "mempool consumption" in 2 steps:
- Collect: we traverse the mempool using the most optimized approach, and collect in one big vector all the eligible transactions
- Sorting: we apply an algorithm, for ordering and packing transactions in blocks / microblocks, that can be smarter thanks to the knowledge of the whole load.
Thoughts?

Reply:
This makes sense to me - @kantai @jcnelson any thoughts?

Reply:
I'd want to be very careful about that, because a naive implementation could easily turn into a DDoS vector.  In general, I'd like to avoid loading _O(n)_ state into RAM if the network attacker controls _n_, as would be the case here.

I think it would be safer to structure the mempool's indexes so the miner could do something like "select each group of transactions that form a nonce chain that can potentially be attached to the chain state, and order the groups by total transaction fees in descending order."  This would let us stream chained transactions into RAM through an iterator, which takes only _O(1)_ RAM.  Then, the node would pick up and mine chained transactions that are contiguous with the current chain state, and would prioritize the transaction _groups_ with the highest total payouts (which also gives us the desirable property that a single very-high-fee transaction can get prioritized over a large chain with a lower total fee).  If the node runs out of block budget while mining, it just moves onto the next-highest-fee group and tries to go from there.

What do y'all think?

Reply:
But, if you are looking for refactoring advice, one thing you might consider doing is turning the code that finds the ancestor consensus hash and block hash from the next block height (by using the chain tip's consensus hash and block hash, and the indexed connection to the sortition DB) into its own method.  This would be the code in lines 434-515.  Up to you though.

Reply:
> I'd want to be very careful about that, because a naive implementation could easily turn into a DDoS vector. In general, I'd like to avoid loading O(n) state into RAM if the network attacker controls n, as would be the case here.

I definitely agree with this. However, I don't think that I agree with the next paragraph. Thinking ahead, I don't think we should try to make too opinionated a choice about mempool iteration, but rather provide an interface for the miner to iterate however they want -- providing different options for "sorting" does that without trying to make miner's choices for them (i.e., trying to compute fee-weighted transaction sets).

In thinking about this more, I think that this PR doesn't really solve issue #2393: providing the ability to iterate transactions in nonce-order, rather than walk-order. This PR just reverses the walk which we expect to help alleviate the problem. If we wanted to apply some other kind of sort in the traversal, I'm not sure how to do that iteratively with the way the mempool is currently structured -- I suspect we'd need to do something like construct the full set of blocks in the current fork (up to the mempool age limit) and use that to filter the SQLite query. I do not know how well SQLite would perform with a large set query like that.

Reply:
> Thinking ahead, I don't think we should try to make too opinionated a choice about mempool iteration, but rather provide an interface for the miner to iterate however they want -- providing different options for "sorting" does that without trying to make miner's choices for them (i.e., trying to compute fee-weighted transaction sets).

Yes, agreed!  But, the issue this PR addresses _is_ an opinionated walk order (it just happens to be more sensible than the current walk order).  Are you saying you think that providing an interface for walk orders should be done in this PR?

> I suspect we'd need to do something like construct the full set of blocks in the current fork (up to the mempool age limit) and use that to filter the SQLite query. I do not know how well SQLite would perform with a large set query like that.

What might a "transaction selector" need?  I think it really just needs the following API functions:

* Get the "best" chain tip the Stacks node knows about.  This is _O(1)_.
* Get all block header hashes (i.e. consensus hashes and block hashes) on the fork from which a given "best" chain tip descends, down to the lowest block we keep transactions for.  This is _O(1)_ time and space to gather (_O(n log n)_ for _n_ blocks if we're counting the index searches, but _n_ is _O(1)_ due to the garbage collection limit).
* Get the newest timestamp of a transaction on a given block header hash history (this is at worst a traversal of the header hash history, but that's still _O(1)_ time and space).
* Find all transaction IDs and fees in a chain tip whose transactions have a given origin address, a nonce in a given range, and a "min-age" timestamp.  This is _O(1) time and space to gather (_O(n log n)_ for _n_ transactions from the same origin, including index searches, but _n_ is _O(1)_ due to the chaining depth limit).

Once you have an interface for these things, I think you can build any sort of mempool walker logic.  Moreover, you could easily expose this via the RPC interface, so a downstream miner program could just iterate through transactions it is interested in (each RPC call is _O(1)_ time and space, and also, the node could gate or rate-limit them if they became too burdensome).  Maybe there's room here to further constrain the kinds of transactions you'd want to expose -- e.g. filtering transactions by type variant or by minimum fee or maximum estimated size or cost, and so on.

I don't think the API calls need to be executed within a single Sqlite transaction -- I think you can get the necessary concurrency control in other ways.  You could guard against an intermittent chain reorg changing your view of the mempool by just checking your best chain tip before and after you produce a block (i.e. optimistic concurrency control), and you could guard against new transactions arriving and potentially invalidating your view by selecting transactions no newer than a given timestamp (effectively giving you read snapshot consistency).  In order for the latter to work correctly, we'd need to also alter the RBF logic in the mempool so it doesn't actually replace the old transaction, but just stores both the old and new transactions (existing code paths that query mempool transactions would be updated to always just take the newest version of a transaction, given an origin and a nonce).  This way, an RBF won't invalidate your mempool view.

What do y'all think?